Support for Checkable list items

mattiasflodin opened the pull request:
When ListView has a choiceMode other than none, it can graphically
display which item(s) are currently selected, if (and only if) the list
items implement the Checkable interface. With DSLV, even if list items
implement this interface, it is hidden by the list-item wrapper used
(DragSortItemView).

With this change, DragSortItemView implements the Checkable interface
and forwards it to the wrapped item view if it implements Checkable,
allowing DSLV to function properly in all choice modes.

I considered a different design where we have a
CheckableDragSortItemView that inherits from DragSortItemView. DSLV
would then instantiate different variants depending on what is needed.
However, the only reason I could find for doing this would be to avoid
the few cycles spent on instanceof checks, and the original ListView
code does not seem to have any concerns about doing this (see
ListView.setupChild).

Commit:
Support for Checkable list items

When ListView has a choiceMode other than none, it can graphically
display which item(s) are currently selected, if (and only if) the list
items implement the Checkable interface. With DSLV, even if list items
implement this interface, it is hidden by the list-item wrapper used
(DragSortItemView).

With this change, DragSortItemView implements the Checkable interface
and forwards it to the wrapped item view if it implements Checkable,
allowing DSLV to function properly in all choice modes.

I considered a different design where we have a
CheckableDragSortItemView that inherits from DragSortItemView. DSLV
would then instantiate different variants depending on what is needed.
However, the only reason I could find for doing this would be to avoid
the few cycles spent on instanceof checks, and the original ListView
code does not seem to have any concerns about doing this (see
ListView.setupChild).

bauerca commented:
Awesome! Thanks. You know what would be awesomer? If you could modify one of the examples (or add a new one, preferably) in the demo project that tests this, and include it in this pull request. Cheers!

mattiasflodin commented:
Sure, I'll see what I can do.

imminent commented:
Hi @mattiasflodin nice addition, I believe. Have you seen `CheckableLinearLayout`? It might be better to implement the `Checkable` interface like this:

``` java
@Override
protected void onFinishInflate() {
    super.onFinishInflate();
    // find checkable child
    final int childCount = getChildCount();
    for (int i = 0; i < childCount; ++i) {
        final View view = getChildAt(i);
    if (view instanceof Checkable)
        _checkbox = (Checkable) view;
    }       
}

@Override 
public boolean isChecked() { 
    return _checkbox != null ? _checkbox.isChecked() : false; 
}

@Override 
public void setChecked(boolean checked) {
    if (_checkbox != null) 
        _checkbox.setChecked(checked);
}

@Override 
public void toggle() { 
   if (_checkbox != null)
       _checkbox.toggle();
} 
```

Something like that, that way you are just checking against `null` and the location of the checkable item in the `ViewGroup` is more flexible, since it looks through the children for the item. You can of course restrict that to only the immediate child if that is preferred.

With multiple choice, would multiple choice drag also work? Look at the Gmail app, for example. The Drag and Drop guide recommends that you combine the items into a single item when dragging multiple items. I think it really would only be used to move a set of items to something-like-a-folder or to swipe-to-delete them. So it might be fine with the current DSLV to ignore checked state when dragging.

mattiasflodin commented:
@imminent I see what you mean. I don't think it would be correct to search the child hierarchy for checkable items because that would diverge from the original ListView behavior, which is to only check the one immediate child (and there must always be only one child at that level in the hierarchy).

Checking for null might be preferable though. But is it guaranteed that onFinishInflate is always called? Aren't there situations when the layout is not inflated from an xml resource? I suppose there could be a state variable to track if we
- know that it's a Checkable child,
- know that it is not a Checkable child, or
- do not know yet (but will find out the first time we need to).

There's a balance between simplicity and functionality to maintain here though and I'm not sure if this really adds enough value (i.e. causes a noticable speed difference for the user).

Multiple choice drag sounds nice but it's beyond my scope right now, I have a project of my own that I wish to focus on. Either way, things like these probably turn out better if they are made by someone who actually needs them for a particular task.

imminent commented:
I believe the code comes from how the CheckableTextView is implemented,
might need to be confirmed. The multiple drag was a comment for Carl :)

bauerca commented:
I think we should leave the commit as Mattias made it, for his reasons listed above, which make sense to me.

Multiple choice drag (I assume you mean dragging many items at once) sounds like a bear to implement and get right (I'm not even sure what "right" means); so far, there have been no requests for it.

bauerca commented:
@mattiasflodin On second thought, I think we _should_ implement the CheckableDragSortItemView. Here's why:

DSLV passes wrapped items to ListView. If the wrapped item is always an instance of Checkable, then the setChecked, isChecked, etc. code in ListView.setupChild will always be run because `child instanceof Checkable` will always be true. This behavior is undesirable, obviously, if the wrapped item is not really a Checkable.

Shouldn't be much extra work, just a instanceof check in DSLV's AdapterWrapper.getView

mattiasflodin commented:
@bauerca I'm not really sure how to implement that. It's a catch-22: mAdapter.getView requires me to pass the parent view as an argument, but I don't know which parent view to create before I know the return value of mAdapter.getView. Any ideas? I've sampled a couple of the getView implementations in the framework and none of them seem to use the parent argument, but I guess that's moot. I honestly don't understand why Google put it there at all.

bauerca commented:
One idea:

We pass DSLV.this as parent View to mAdapter.getView(int, View, ViewGroup) **with** the stipulation (in documentation) that in the user's implementation of getView(), if they use LayoutInflater.inflate(int id, ViewGroup root, boolean attach), they **must** pass attach=false.

2nd idea (bomb-proof, but hacky, gross, and redundant):

Make a plain DragSortItemView, pass it to mAdapter.getView, check instanceof, then call mAdapter.getView again (with the result of the first call passed in as convertView, of course) with the correct DSItemView subtype passed in as parent.

Thoughts please...

bauerca commented:
I kinda like the first one since it's common to call LayoutInflater.inflate(id, null) when inflating a ListView item.

mattiasflodin commented:
Well... my thoughts are that this is premature optimization based on assumptions, and that the original version will be just fine unless someone can prove there's a measurable problem with it. :)

Yes, DragSortListView.isChecked etc will be called unnecessarily, but all they do if the child is not Checkable is to call instanceof and then return. What's more, they are only called if the user sets the choiceMode to something other than none. It is pretty unlikely he would do that unless he has Checkable list items.

bauerca commented:
Very premature, indeed. And I like the choiceMode argument; overlooked that browsing the source.

Okay I'm game w/ your original commit; Sorry for the runaround. :) Are you still going to update with an example?

Thanks.
-Carl

mattiasflodin commented:
Yep, working on that.

mattiasflodin commented:
Well, it was a good call to write a simple test case for it. Turns out I have to think a bit more about how to transfer the checked state of items that are moved and who is responsible for it.

(In my "production" app this was not apparent because the selection is updated each time the underlying cursor sends an update notification, which happens on every drag-drop operation).

Commit:
Examples in demo project for single- and multi-choice listview

Commit:
DragSortListView.moveCheckState

Add a method to move the check state of ListView items along with the
items during drag & drop. This is necessary for adapters without stable
IDs, because the ListView has no way of determining how items have moved
when the underlying adapter sends a change notification.

mattiasflodin commented:
I rebased my commits onto the current tip of master, so if you already pulled the old branch into your local repo you may need to force the update (or just delete the branch locally before you pull).

bauerca commented:
Fantastic. Checking it out today.

bauerca commented:
Wow. Really great addition to DSLV. That reordering turned out to be a bear, huh?

Why did you decide to make the dev call moveCheckState in the DropListener? Can't we check `getChoiceMode()` and `getAdapter().hasStableIds()` in `DragSortListView.dropFloatView()`?

I will merge this request locally, increment version numbers, update README, and stuff. I have other new features that might be bundled with this.

mattiasflodin commented:
Thanks, yeah `buildRunList()` was tricky to get right. I found that you need to be pretty careful when making changes to it to avoid breaking anything.

I did consider calling `moveCheckState()` in `dropFloatView()`, but I was afraid it might cause trouble in some cases. If the user does not update the adapter at all or if something more sophisticated than a simple move happens after a drop (such as moving something else along with the item you moved? I don't know) then `moveCheckState` will do the wrong thing.

In my own app, things can move around due to external events (think e.g. Tasker) and I have code set up already to handle arbitrary change notifications and set the correct check state. If `moveCheckState()` always happens in `dropFloatView` then at best it will be redundant, and at worst it will cause an inconsistent check state.

That said, I don't know how much it will be a problem in practice, but it seemed more consistent with the rest of the design to do it in the drop handler. If you feel the other way is better then I'm sure that will work out fine in the real world too.

bauerca commented:
Ugh. We need to handle removals, too. Maybe this is quicker for you to implement?

mattiasflodin commented:
Right, of course we do... I'll take a look at it.

mattiasflodin commented:
Just thought I'd pop in to let you know I haven't forgotten about this, I've just run into a lot of IRL distractions (flu etc in family + Christmas preparations). It may take a while longer before I get to finishing it, but I don't expect there to be much trouble once I get to it.

bauerca commented:
Word. No rush. Glad to hear your family's immune systems are functioning properly.

Commit:
DragSortListView.removeCheckState

Method similar to moveCheckState but for moving check statuses up one step
when a list item was removed.

Commit:
Examples in demo project for checkable listviews with removable items

mattiasflodin commented:
Thanks for the patience. I've pushed a new attempt including a removeCheckState() method.

bauerca commented:
Fantastic.
